caravan.lua

```diff
@@ -27,6 +27,312 @@ specified, then the commands apply to all caravans on the map.
 
 --@ module = true
 
+local gui = require('gui')
+local overlay = require('plugins.overlay')
+local widgets = require('gui.widgets')
+
+trader_selected_state = trader_selected_state or {}
+broker_selected_state = broker_selected_state or {}
+handle_ctrl_click_on_render = handle_ctrl_click_on_render or false
+handle_shift_click_on_render = handle_shift_click_on_render or false
+
+dfhack.onStateChange.caravanOverlay = function(code)
+    if code == SC_WORLD_UNLOADED then
+        trader_selected_state = {}
+        broker_selected_state = {}
+        handle_ctrl_click_on_render = false
+        handle_shift_click_on_render = false
+    end
+end
+
+local GOODFLAG = {
+    UNCONTAINED_UNSELECTED = 0,
+    UNCONTAINED_SELECTED = 1,
+    CONTAINED_UNSELECTED = 2,
+    CONTAINED_SELECTED = 3,
+    CONTAINER_COLLAPSED_UNSELECTED = 4,
+    CONTAINER_COLLAPSED_SELECTED = 5,
+}
+
+function is_trade_overlay_active()
+    return df.global.game.main_interface.trade.open
+end
```

Comment:
this is still required for now, but soon I hope to be using your new focus string logic instead. the `viewscreens` attribute will become `dwarfmode@focus_string` and the overlay framework will do the matching.

Reply:
once DFHack/dfhack#2825 is merged, then this can be expressed as a focus string in the `viewscreens` attribute and the `is_trade_overlay_active` function can be removed.